### PR TITLE
security(auth): harden auth posture (fix open-redirect & info-leak)

### PR DESCRIPTION
**Pull request by hamza-hafeez82**
# Security: Production Log Sanitization & Auth Logic Tests

##  Summary
After doing a manual review of the backend security posture, I’ve put together this PR to address some production logging risks and establish a solid testing foundation for your authentication logic. The goal is to make InsForge a bit more "production-hardened" without getting in the way of the great developer experience (DX) the team has built.

---

### Production Log Sanitization (Fixing Info Leakage)
Currently, the Winston logger sends full error stack traces to the logs in every environment. While this is helpful during development, it's a security risk in production.
*   **The Problem:** Stack traces leak internal file paths and module structures, which can serve as a "blueprint" for attackers during reconnaissance (**OWASP A09:2021**).
*   **The Fix:** I’ve updated the logger to only include the `stack` when `NODE_ENV` is NOT `production`. You still get the error message and metadata in prod, but the internal system details stay hidden.

---

### New Unit Test Suite for Auth Config
We have some great logic for redirect wildcards and normalization, so i wanted to make sure we have regression protection for it.
*   **What's Tested:** I’ve added `tests/unit/auth-config.service.test.ts` to cover exact matches, subdomain wildcards (`*.domain.com`), and various DB edge cases (like `null` whitelists).
*   **Keeping it Permissive:** Per the team's feedback on the intended design, the tests explicitly verify that the system remains permissive when the whitelist is empty. This ensures we don't accidentally introduce breaking changes in the future while still documenting and protecting the current behavior.

---

## Validation
I've made sure everything is clean before submitting:
*   Ran `turbo run typecheck lint` across the backend.
*   All 6 new unit tests for `AuthConfigService` are passing successfully.

## References
*   **OWASP A09:2021** , Security Logging and Monitoring Failures (addressing stack trace exposure).
*   **OWASP A07:2021** , Identification and Authentication Failures (improving test coverage for auth logic).

### CODE REVIEW

**Comment by tonychang04** [comment]
this is good change but i think it will break the users who haven't setup the redirect url. when they upgraded the previous version to the new version. What we can do is recommending they should configure redirect url

**Comment by hmza-hb** [comment]
> this is good change but i think it will break the users who haven't setup the redirect url. when they upgraded the previous version to the new version. What we can do is recommending they should configure redirect url

Yeah, that’s a fair point actually. My thinking was mainly around avoiding permissive defaults on fresh or partially configured deployments, but I agree this could break auth flows for users upgrading from older versions without redirect URLs configured yet.

A warning + migration path might make more sense first before enforcing it strictly by default. I’m happy to adjust the implementation around that.

**Comment by tonychang04** [comment]
thanks for you chage on this. I think for recommendation + warning path 
Screenshot 2026-05-07 at 10 24 24
 should belong to our newer version for backend advisor?  cc @jwfing  what do you think, whichi is out of scope for the opensource

**Comment by hmza-hb** [comment]
> thanks for you chage on this. I think for recommendation + warning path Screenshot 2026-05-07 at 10 24 24 should belong to our newer version for backend advisor? cc @jwfing what do you think, whichi is out of scope for the opensource

Thanks for the feedback, @tonychang04 

I've just updated the implementation to follow the recommendation + warning path as we discussed.

What's changed:

Permissive with Warning: If the redirect whitelist is empty or null, the system now logs a clear SECURITY WARNING but remains permissive. This ensures zero friction for users upgrading from older versions while giving them the necessary visibility into the open-redirect risk.
Edge Case Coverage: I've added a unit test to cover the null scenario (database default) as suggested by CodeRabbit to ensure the logic is robust.
Regarding the Backend Advisor, that looks like a fantastic tool! I agree that systemic security checks belong there in the long run. In the meantime, this log-based warning serves as a lightweight 'first line of defense' for the open-source community.

Looking forward to your thoughts on the revised logic!

File: backend/src/services/auth/auth-config.service.ts [hunk 1]
```diff
@@ -276,7 +276,13 @@ export class AuthConfigService {
...
     const allowedRedirectUrls = config.allowedRedirectUrls;
 
+    // Security: If no whitelist is configured, log a warning but remain permissive for now
+    // to avoid breaking existing users during upgrades. We recommend admins configure
+    // this whitelist as soon as possible to prevent open redirect attacks.
     if (!allowedRedirectUrls || allowedRedirectUrls.length === 0) {
+      logger.warn(
+        'SECURITY WARNING: No allowed redirect URLs are configured. Authentication redirects will be permissive, which allows open redirect attacks. Please configure your allowed redirect URLs in the Authentication configuration dashboard.'
+      );
       return true;
```
[the remark this answers is not in the document]
Comment (hmza-hb) [reply backend/src/services/auth/auth-config.service.ts]: I appreciate the automated scan flagging this. There is a deliberate tradeoff here between Security and Backward Compatibility.

As @tonychang04 pointed out, a strict default-deny would break authentication for a significant number of existing InsForge users who haven't yet configured their redirect whitelists.

The 'Permissive with Warning' approach is an intentional bridge:

It fixes the information leak (SEC-009) completely.
It provides immediate visibility into the Open Redirect risk (SEC-002) via the security warning in logs, which previously was silent.
It prevents production outages for users upgrading from older versions.
Moving to a strict default-deny is a great candidate for a future major release or the 'Backend Advisor' tool discussed above. For this PR, we are prioritizing a safe migration path while improving the security baseline.

File: backend/src/services/auth/auth-config.service.ts [hunk 2]
```diff
@@ -276,7 +276,13 @@ export class AuthConfigService {
...
     }
```
[the remark this answers is not in the document]
Comment (hmza-hb) [reply backend/src/services/auth/auth-config.service.ts:284]: I appreciate the automated scan flagging this. There is a deliberate tradeoff here between Security and Backward Compatibility.

As @tonychang04 pointed out, a strict default-deny would break authentication for a significant number of existing InsForge users who haven't yet configured their redirect whitelists.

The 'Permissive with Warning' approach is an intentional bridge:

It fixes the information leak (SEC-009) completely.
It provides immediate visibility into the Open Redirect risk (SEC-002) via the security warning in logs, which previously was silent.
It prevents production outages for users upgrading from older versions.
Moving to a strict default-deny is a great candidate for a future major release or the 'Backend Advisor' tool discussed above. For this PR, we are prioritizing a safe migration path while improving the security baseline.

**Comment by Fermionic-Lyu** [comment]
The allowed redirect urls is our **intended** design. Please don't change that.

**Comment by hmza-hb** [comment]
> The allowed redirect urls is our **intended** design. Please don't change that.

Hi @Fermionic-Lyu, thanks for jumping in!

Just to be sure i'm on the right track: is the concern with the log warning i added for empty whitelists, or is there a specific part of the matching logic you'd like me to keep exactly as is?

To clarify, the current version of the PR does not change the core matching behavior (wildcards and normalization remain the same). It only adds a log warning when the whitelist is empty to address a security scanner finding, while keeping the behavior permissive.

I'm happy to revert the warning if you feel the current 'silent' permissive behavior is the intended design for the open-source version!

**Comment by Fermionic-Lyu** [comment]
> > The allowed redirect urls is our **intended** design. Please don't change that.
> 
> Hi @Fermionic-Lyu, thanks for jumping in!
> 
> Just to be sure i'm on the right track: is the concern with the log warning i added for empty whitelists, or is there a specific part of the matching logic you'd like me to keep exactly as is?
> 
> To clarify, the current version of the PR does not change the core matching behavior (wildcards and normalization remain the same). It only adds a log warning when the whitelist is empty to address a security scanner finding, while keeping the behavior permissive.
> 
> I'm happy to revert the warning if you feel the current 'silent' permissive behavior is the intended design for the open-source version!

Yes, we allow empty allowed list to be permissive to improve DX and lower development frictions. We do have a warning badge to inform the developers that for production env, you should configure allowed redirect urls specifically for better security.

At this moment, let's not change anything around that. You might also need to update your PR description so our review agents don't go crazy on criticizing.

For the winston logger changes, I think it is good but let's get a confirm from @jwfing as well

**Comment by hmza-hb** [comment]
> > > The allowed redirect urls is our **intended** design. Please don't change that.
> > 
> > 
> > Hi @Fermionic-Lyu, thanks for jumping in!
> > Just to be sure i'm on the right track: is the concern with the log warning i added for empty whitelists, or is there a specific part of the matching logic you'd like me to keep exactly as is?
> > To clarify, the current version of the PR does not change the core matching behavior (wildcards and normalization remain the same). It only adds a log warning when the whitelist is empty to address a security scanner finding, while keeping the behavior permissive.
> > I'm happy to revert the warning if you feel the current 'silent' permissive behavior is the intended design for the open-source version!
> 
> Yes, we allow empty allowed list to be permissive to improve DX and lower development frictions. We do have a warning badge to inform the developers that for production env, you should configure allowed redirect urls specifically for better security.
> 
> At this moment, let's not change anything around that. You might also need to update your PR description so our review agents don't go crazy on criticizing.
> 
> For the winston logger changes, I think it is good but let's get a confirm from @jwfing as well

Thanks for the context, @Fermionic-Lyu! That makes perfect sense.

I have just pushed an update that aligns with your vision:

Reverted Redirect Logic: I've removed the hardening and the log warning for the redirect whitelist. It's now back to the original 'silent permissive' behavior.
Updated Tests: I kept the new unit tests but updated them to assert and protect this intended permissive behavior. This way, we have regression protection for the design you've established.
Updated PR Description: I've pivoted the PR description to focus on the Winston logger fix and the new test suite, so the review agents can focus on the right things.
Looking forward to the confirm from @jwfing on the logger changes!

**Review by jwfing — APPROVED** [review]
